File: advent_of_code_2024/day06/Main.hs
module Main where

import Data.Foldable (traverse_)
import Data.Map.Strict qualified as M
import Data.Set qualified as S
import Linear.V2 (V2 (..))

type Coord = V2 Int
type Map = M.Map Coord Char
data Dir = U | D | L | R deriving (Eq, Ord)
data Guard a = Guard {pos :: Coord, dir :: Dir, visited :: S.Set a} deriving (Eq)

parse :: String -> Map
parse input =
    M.fromList
        [ (V2 j i, c)
        | (j, l) <- zip [0 ..] $ reverse $ lines input
        , (i, c) <- zip [0 ..] l
        ]

start :: (Coord -> Dir -> a) -> Map -> (Map, Guard a)
start f m = (m', Guard p U (S.singleton $ f p U))
  where
    (p, _) = M.findMin $ M.filter (== '^') m
    m' = M.update (const $ Just '.') p m

turnRight :: Dir -> Dir
turnRight = \case U -> R; D -> L; L -> U; R -> D

toStep :: Dir -> V2 Int
toStep = \case U -> V2 1 0; D -> V2 (-1) 0; L -> V2 0 (-1); R -> V2 0 1

next :: (Ord a) => (Coord -> Dir -> a) -> Map -> Guard a -> Guard a
next f m Guard{pos, dir, visited} = Guard pos' dir' visited'
  where
    is xy c = m M.!? xy == Just c
    ahead = pos + toStep dir
    blocked = ahead `is` '#'
    done = not $ M.member ahead m
    dir' = if blocked then turnRight dir else dir
    pos' = if blocked || done then pos else ahead
    visited' = if blocked || done then visited else S.insert (f pos' dir') visited

part1 :: Map -> Int
part1 m = length . visited $ go g
  where
    (m', g) = start const m
    go x = let y = next const m' x in if x == y then x else go y

part2 :: Map -> Int
part2 m = length . filter ((&&) <$> loops <*> (/= p0)) $ M.keys m'
  where
    (m', g0) = start (,) m
    p0 = pos g0
    loops p = go g0
      where
        go g = let g' = next (,) m'' g in (g' /= g) && ((pos g', dir g') `S.member` visited g || go g')
        m'' = M.update (const $ Just '#') p m'

main :: IO ()
main = do
    m <- parse <$> readFile "input.txt"
    traverse_ (print . ($ m)) [part1, part2]

File: advent_of_code_2024/day06/src/main.rs
use eyre::{eyre, Result, WrapErr};
use itertools::Itertools;
use rayon::prelude::*;
use std::{collections::BTreeSet, fs};

#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
struct Coord(u8, u8);

#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
enum Dir {
    U,
    D,
    L,
    R,
}

impl Dir {
    fn turn_right(self) -> Self {
        match self {
            Self::U => Self::R,
            Self::D => Self::L,
            Self::L => Self::U,
            Self::R => Self::D,
        }
    }
}

impl Dir {
    fn step(self, c: Coord) -> Coord {
        let (i, j) = (c.0, c.1);
        match self {
            Dir::U => Coord(i + 1, j),
            Dir::D => Coord(i.wrapping_sub(1), j),
            Dir::L => Coord(i, j.wrapping_sub(1)),
            Dir::R => Coord(i, j + 1),
        }
    }
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
struct Guard {
    pos: Coord,
    dir: Dir,
}

trait Trail: Ord + Sized {
    fn f(g: Guard) -> Self;
    fn stop_early(&self, visited: &BTreeSet<Self>) -> bool;
}
impl Trail for Coord {
    fn f(g: Guard) -> Self {
        g.pos
    }
    fn stop_early(&self, _visited: &BTreeSet<Self>) -> bool {
        false
    }
}
impl Trail for (Coord, Dir) {
    fn f(g: Guard) -> Self {
        (g.pos, g.dir)
    }
    fn stop_early(&self, visited: &BTreeSet<Self>) -> bool {
        visited.contains(self)
    }
}

enum Walk {
    EarlyStop(usize),
    OutOfBounds(usize),
    Bail,
}

fn ix(n: u8, i: u8, j: u8) -> usize {
    n as usize * j as usize + i as usize
}

#[derive(Clone)]
struct Map(u8, Vec<bool>);

impl Map {
    fn contains(&self, p: Coord) -> bool {
        p.0 < self.0 && p.1 < self.0
    }
    fn get(&self, p: Coord) -> bool {
        self.1[ix(self.0, p.0, p.1)]
    }
    fn set(&mut self, p: Coord, c: bool) {
        self.1[ix(self.0, p.0, p.1)] = c;
    }
    fn parse(n: usize, input: &str) -> Result<(Self, Guard)> {
        let mut data = vec![false; n * n];
        let n: u8 = n.try_into().wrap_err("n too big")?;
        let mut pos = Coord(255, 255);
        for (i, line) in input.lines().rev().enumerate() {
            for (j, c) in line.chars().enumerate() {
                let i: u8 = i.try_into().wrap_err("i too big")?;
                let j: u8 = j.try_into().wrap_err("i too big")?;
                data[ix(n, i, j)] = c == '#';
                if c == '^' {
                    pos = Coord(i, j);
                }
            }
        }
        if pos == Coord(255, 255) {
            Err(eyre!("No ^ found."))
        } else {
            Ok((Self(n, data), Guard { pos, dir: Dir::U }))
        }
    }
    fn go<T: Trail>(&self, mut g: Guard) -> Walk {
        let mut visited = BTreeSet::from([T::f(g)]);
        let mut n = 0;
        while n < self.1.len() {
            let ahead = g.dir.step(g.pos);
            if self.contains(ahead) {
                if self.get(ahead) {
                    g.dir = g.dir.turn_right();
                } else {
                    g.pos = ahead;
                }
                let trail = T::f(g);
                if trail.stop_early(&visited) {
                    return Walk::EarlyStop(visited.len());
                }
                visited.insert(trail);
            } else {
                return Walk::OutOfBounds(visited.len());
            }
            n += 1;
        }
        Walk::Bail
    }
}

fn part1(m: &Map, g: Guard) -> Result<usize> {
    match m.go::<Coord>(g) {
        Walk::EarlyStop(n) | Walk::OutOfBounds(n) => Ok(n),
        Walk::Bail => Err(eyre!("Bigger than map's size")),
    }
}

fn part2(m: &Map, g: Guard) -> usize {
    (0..m.0)
        .cartesian_product(0..m.0)
        .filter(|&ij| ij != (g.pos.0, g.pos.1))
        .par_bridge()
        .filter(|&(i, j)| {
            let mut m_ = m.clone();
            m_.set(Coord(i, j), true);
            matches!(m_.go::<(Coord, Dir)>(g), Walk::EarlyStop(_))
        })
        .count()
}

fn main() -> Result<()> {
    let input = fs::read_to_string("input.txt").wrap_err("Unable to read input file.")?;
    let n = input.chars().filter(|c| *c == '\n').count();
    let (m, g) = Map::parse(1 + n, &input)?;
    println!("{}", part1(&m, g)?);
    println!("{}", part2(&m, g));
    Ok(())
}
